ocaml/file_formats/cms_format.ml

```diff
@@ -71,37 +73,41 @@ let uid_tables_of_binary_annots binary_annots =
     !toplevel_attributes;
   Cmt_format.iter_declarations binary_annots
     ~f:(fun uid decl ->
-      let loc, attrs =
+      let loc = Typedtree.loc_of_decl ~uid decl in
+      let attrs =
         match decl with
-        | Value v -> v.val_loc, v.val_attributes
-        | Value_binding v -> v.vb_loc, v.vb_attributes
-        | Type v -> v.typ_loc, v.typ_attributes
-        | Constructor v -> v.cd_loc, v.cd_attributes
-        | Extension_constructor v -> v.ext_loc, v.ext_attributes
-        | Label v -> v.ld_loc, v.ld_attributes
-        | Module v -> v.md_loc, v.md_attributes
-        | Module_substitution v -> v.ms_loc, v.ms_attributes
-        | Module_binding v -> v.mb_loc, v.mb_attributes
-        | Module_type v -> v.mtd_loc, v.mtd_attributes
-        | Class v -> v.ci_loc, v.ci_attributes
-        | Class_type v -> v.ci_loc, v.ci_attributes
+        | Value v -> v.val_attributes
+        | Value_binding v -> v.vb_attributes
+        | Type v -> v.typ_attributes
+        | Constructor v -> v.cd_attributes
+        | Extension_constructor v -> v.ext_attributes
+        | Label v -> v.ld_attributes
+        | Module v -> v.md_attributes
+        | Module_substitution v -> v.ms_attributes
+        | Module_binding v -> v.mb_attributes
+        | Module_type v -> v.mtd_attributes
+        | Class v -> v.ci_attributes
+        | Class_type v -> v.ci_attributes
       in
       Types.Uid.Tbl.add cms_uid_to_loc uid loc;
       Types.Uid.Tbl.add cms_uid_to_attributes uid attrs
     );
   cms_uid_to_loc, cms_uid_to_attributes
 
-let save_cms filename modname binary_annots sourcefile shape =
+let save_cms filename modname binary_annots sourcefile initial_env shape =
   if (!Clflags.binary_annotations_cms && not !Clflags.print_types) then begin
     Misc.output_to_file_via_temporary
        ~mode:[Open_binary] filename
        (fun _temp_file_name oc ->
         let source_digest = Option.map Digest.file sourcefile in
-        let cms_ident_occurrences =
+        let cms_ident_occurrences, cms_initial_env =
           if !Clflags.store_occurrences then
-            Cmt_format.index_occurrences binary_annots
+            let cms_ident_occurrences = Cmt_format.index_occurrences binary_annots in
+            let cms_initial_env = if Cmt_format.need_to_clear_env
+              then Env.keep_only_summary initial_env else initial_env in
+            cms_ident_occurrences, Some cms_initial_env
           else
-            [| |]
+            [| |], None
         in
         let cms_uid_to_loc, cms_uid_to_attributes =
```

Comment:
This `cms_uid_to_loc` is generated here, only to be converted to an array just below. Is it possible to create the array directly, without going via a table first?

Reply:
Do you know whether two `Typedtree.item_declaration`s can define the same `Types.Uid.t`? My understanding is no, in which case this is very doable.

If the answer is yes, then we would need to be slightly careful here about duplicates. We'd either need to dedup here (which would be difficult without building a table) or rely on Merlin to handle duplicates properly.

Reply:
It appears that there can indeed be duplicates. For now, I'm going to continue encoding it as a hashtbl